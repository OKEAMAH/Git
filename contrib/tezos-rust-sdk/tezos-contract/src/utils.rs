use tezos_michelson::michelson::metadata::TypeFieldMetadata;

pub trait AnyAnnotationValue {
    fn any_annotation_value(&self) -> Option<&str>;
}

impl AnyAnnotationValue for TypeFieldMetadata {
    fn any_annotation_value(&self) -> Option<&str> {
        self.field_name().as_ref().map_or(
            self.type_name()
                .as_ref()
                .map(|annotation| annotation.value_without_prefix()),
            |annotation| Some(annotation.value_without_prefix()),
        )
    }
}
