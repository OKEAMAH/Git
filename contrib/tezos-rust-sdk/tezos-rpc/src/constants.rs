/// An alias that points to the latest block.
pub const BLOCK_HEAD_ALIAS: &str = "head";
/// An alias that points to the first block.
pub const BLOCK_GENESIS_ALIAS: &str = "genesis";
