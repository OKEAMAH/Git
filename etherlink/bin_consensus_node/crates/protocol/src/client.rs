// SPDX-FileCopyrightText: 2024 TriliTech <contact@trili.tech>
//
// SPDX-License-Identifier: MIT

//! Local protocol client.

use async_trait::async_trait;
use dsn_core::api::{ApiError, DsnApi};
use dsn_core::{
    storage::StorageBackend,
    types::{PreBlock, PreBlockHeader, Transaction},
};
use tokio::sync::{broadcast, mpsc};

use crate::error::ProtocolError;
use crate::{state::ProtocolState, ProtocolClient};

impl<S: StorageBackend> ProtocolClient<S> {
    pub fn new(
        state: ProtocolState<S>,
        rx_pre_blocks: broadcast::Receiver<PreBlock>,
        tx_mempool: mpsc::Sender<Transaction>,
    ) -> Self {
        Self {
            state,
            rx_pre_blocks,
            tx_mempool,
        }
    }
}

impl<S: StorageBackend> Clone for ProtocolClient<S> {
    fn clone(&self) -> Self {
        Self {
            state: self.state.clone(),
            rx_pre_blocks: self.rx_pre_blocks.resubscribe(),
            tx_mempool: self.tx_mempool.clone(),
        }
    }
}

#[async_trait]
impl<S: StorageBackend> DsnApi for ProtocolClient<S> {
    async fn submit_transaction(&self, transaction: Transaction) -> Result<(), ApiError> {
        self.tx_mempool
            .send(transaction)
            .await
            .map_err(Into::<ProtocolError>::into)
            .map_err(Into::into)
    }

    async fn get_pre_blocks_head(&self) -> Result<PreBlockHeader, ApiError> {
        match self.state.get_head() {
            Ok(value) => Ok(value),
            Err(ProtocolError::MissingPreBlockHead) => Err(ApiError::ItemNotFound),
            Err(err) => Err(err.into()),
        }
    }

    async fn get_pre_blocks(
        &self,
        from_id: u64,
        max_count: usize,
    ) -> Result<Vec<PreBlock>, ApiError> {
        self.state
            .get_pre_blocks(from_id, max_count)
            .map_err(Into::into)
    }

    async fn next_pre_block(&mut self) -> Result<PreBlock, ApiError> {
        self.rx_pre_blocks
            .recv()
            .await
            .map_err(|_| ApiError::ShutdownInProgress)
    }

    async fn clear_queue(&mut self) -> Result<(), ApiError> {
        self.rx_pre_blocks = self.rx_pre_blocks.resubscribe();
        Ok(())
    }
}
