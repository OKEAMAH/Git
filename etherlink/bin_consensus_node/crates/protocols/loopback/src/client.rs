// SPDX-FileCopyrightText: 2024 TriliTech <contact@trili.tech>
//
// SPDX-License-Identifier: MIT

use async_trait::async_trait;
use dsn_core::traits::{ApiError, PreBlocksApi, TransactionsApi};
use dsn_core::{
    storage::StorageBackend,
    types::{PreBlock, PreBlockHeader, Transaction},
};
use tokio::sync::{broadcast, mpsc};

use crate::error::LoopbackError;
use crate::{state::LoopbackState, LoopbackClient};

impl<S: StorageBackend> LoopbackClient<S> {
    pub fn new(
        state: LoopbackState<S>,
        rx_pre_blocks: broadcast::Receiver<PreBlock>,
        tx_mempool: mpsc::Sender<Transaction>,
    ) -> Self {
        Self {
            state,
            rx_pre_blocks,
            tx_mempool,
        }
    }
}

impl<S: StorageBackend> Clone for LoopbackClient<S> {
    fn clone(&self) -> Self {
        Self {
            state: self.state.clone(),
            rx_pre_blocks: self.rx_pre_blocks.resubscribe(),
            tx_mempool: self.tx_mempool.clone(),
        }
    }
}

#[async_trait]
impl<S: StorageBackend> TransactionsApi for LoopbackClient<S> {
    async fn submit_transaction(&self, transaction: Transaction) -> Result<(), ApiError> {
        self.tx_mempool
            .send(transaction)
            .await
            .map_err(Into::<LoopbackError>::into)
            .map_err(Into::into)
    }
}

#[async_trait]
impl<S: StorageBackend> PreBlocksApi for LoopbackClient<S> {
    async fn get_pre_blocks_head(&self) -> Result<PreBlockHeader, ApiError> {
        match self.state.get_head() {
            Ok(value) => Ok(value),
            Err(LoopbackError::MissingPreBlockHead) => Err(ApiError::NotFound),
            Err(err) => Err(err.into()),
        }
    }

    async fn get_pre_blocks(
        &self,
        from_id: u64,
        max_count: usize,
    ) -> Result<Vec<PreBlock>, ApiError> {
        self.state
            .get_pre_blocks(from_id, max_count)
            .map_err(Into::into)
    }

    async fn next_pre_block(&mut self) -> Result<PreBlock, ApiError> {
        self.rx_pre_blocks
            .recv()
            .await
            .map_err(|_| ApiError::Shutdown)
    }

    async fn clear_queue(&mut self) -> Result<(), ApiError> {
        self.rx_pre_blocks = self.rx_pre_blocks.resubscribe();
        Ok(())
    }
}
