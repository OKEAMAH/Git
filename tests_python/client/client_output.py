"""Structured representation of client output."""
from typing import List
import re
import json
# TODO This is incomplete. Add additional attributes and result classes as
#      they are needed


class InvalidClientOutput(Exception):
    """Raised when client output couldn't be parsed."""

    def __init__(self, client_output: str):
        super().__init__(self)
        self.client_output = client_output


class EndorseResult:
    """Result of a 'endorse for' operation."""

    def __init__(self, client_output: str):
        pattern = r"Operation hash is '?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.operation_hash = match.groups()[0]


class TransferResult:
    """Result of a 'transfer' operation."""

    def __init__(self, client_output: str):
        self.client_output = client_output
        pattern = r"Operation hash is '?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.operation_hash = match.groups()[0]
        pattern = r"--branch ?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.branch_hash = match.groups()[0]


class GetReceiptResult:
    """Result of 'get receipt' operation.

    If operation wasn't found, 'black_hash' is set to None.
    """

    def __init__(self, client_output: str):
        if client_output == "Couldn't find operation\n":
            self.block_hash = None
            return
        pattern = r"Operation found in block: ?(\w*) "
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.block_hash = match.groups()[0]


class GetAddressesResult:
    """Result of 'list known addresses' operation.

    """

    def __init__(self, client_output: str):

        pattern = re.compile(r"^(\w+):\s*(\w+).*$", re.MULTILINE)
        self.wallet = dict(re.findall(pattern, client_output))


class RunScriptResult:
    """Result of a 'get script' operation."""

    def __init__(self, client_output: str):
        # read storage output
        pattern = r"(?s)storage\n\s*(.*)\nemitted operations\n"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.storage = match.groups()[0]

        # read map diff output
        self.big_map_diff = []
        pattern = r"big_map diff\n"
        match = re.search(pattern, client_output)
        if match is not None:
            pattern = re.compile(r"  ([^ ].*?)\n")
            for match_diff in pattern.finditer(client_output, match.end(0)):
                self.big_map_diff.append([match_diff.group(1)])


class OriginationResult:
    """Result of an 'originate contract' operation."""

    def __init__(self, client_output: str):
        pattern = r"New contract ?(\w*) originated"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.contract = match.groups()[0]
        pattern = r"Operation hash is '?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput
        self.operation_hash = match.groups()[0]


class SubmitProposalsResult:
    """Result of an 'submit proposals' operation."""

    def __init__(self, client_output: str):
        pattern = r"Operation hash is '?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.operation_hash = match.groups()[0]


class BakeForResult:
    """Result of a 'baker for' operation."""

    def __init__(self, client_output: str):
        pattern = r"Injected block ?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.block_hash = match.groups()[0]


class ActivationResult:
    """Result of 'activate protocol' command"""

    def __init__(self, client_output: str):
        pattern = r"Injected ?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.block_hash = match.groups()[0]


class WaitForResult:
    """Result of a 'wait for' command."""

    def __init__(self, client_output: str):
        pattern = r"Operation found in block: ?(\w*) "
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.block_hash = match.groups()[0]


class HashResult:
    """Result of a 'hash data' command."""

    def __init__(self, client_output: str):

        pattern = r'''Raw packed data: ?(0x[0-9a-f]*)
Script-expression-ID-Hash: ?(\w*)
Raw Script-expression-ID-Hash: ?(\w*)
.*
Raw Sha256 hash: ?(\w*)
Raw Sha512 hash: ?(\w*)'''
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.packed = match.groups()[0]
        self.hash = match.groups()[1]
        self.blake2b = match.groups()[2]
        self.sha256 = match.groups()[3]
        self.sha512 = match.groups()[4]


class SignatureResult:
    """Result of a 'sign bytes' command."""

    def __init__(self, client_output: str):

        pattern = r'Signature: ?(\w*)\n'
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.sig = match.groups()[0]


class SetDelegateResult:
    """Result of a 'set delegate' operation."""

    def __init__(self, client_output: str):
        pattern = r"Operation hash is '?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.operation_hash = match.groups()[0]
        pattern = r"--branch ?(\w*)"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        self.branch_hash = match.groups()[0]


class GetDelegateResult:
    """Result of a 'get delegate' command."""

    def __init__(self, client_output: str):
        if client_output == 'none\n':
            self.delegate = None
        else:
            pattern = r"(\w+)( \(known as (\w+)\))*"
            match = re.search(pattern, client_output)
            if match is None:
                raise InvalidClientOutput(client_output)
            self.address = match.groups()[0]
            self.alias = match.groups()[2]
            self.delegate = self.address


class SignBytesResult:
    """Result of a 'sign bytes' command."""

    def __init__(self, client_output: str):
        pattern = r"Signature: ?(\w+)"
        match = re.search(pattern, client_output[:-1])
        if match is None:
            raise InvalidClientOutput(client_output)
        self.signature = match.groups()[0]


def extract_rpc_answer(client_output: str) -> dict:
    """Convert json client output to a dict representation.

    For some RPC, the answer isn't json, in that case, the client_output
    can be retrieved from the InvalidClientOutput exception"""
    try:
        return json.loads(client_output)
    except json.JSONDecodeError:
        raise InvalidClientOutput(client_output)


def extract_balance(client_output: str) -> float:
    """Extract float balance from the output of 'get_balance' operation."""
    try:
        pattern = r"([\w.]*) êœ©"
        match = re.search(pattern, client_output)
        if match is None:
            raise InvalidClientOutput(client_output)
        return float(match.groups()[0])
    except Exception:
        raise InvalidClientOutput(client_output)


def extract_protocols(client_output: str) -> List[str]:
    """Extract protocol from the output of 'get_protocols' operation."""
    return client_output.split()
