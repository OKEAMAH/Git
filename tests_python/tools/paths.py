import os


def tezos_home():
    # we rely on a fixed directory structure of the tezos repo
    # TEZOS_HOME/tests_python/tools/paths.py
    cur_path = os.path.dirname(os.path.realpath(__file__))
    tezos_home = os.path.dirname(os.path.dirname(cur_path)) + '/'
    assert os.path.isfile(f'{tezos_home}/tests_python/tools/paths.py')
    return tezos_home


# Use environment variable if tests_python are put outside the tezos
# TEZOS_HOME = os.environ.get('TEZOS_HOME')
TEZOS_HOME = tezos_home()
TEZOS_BINARIES = os.environ.get('TEZOS_BINARIES')
