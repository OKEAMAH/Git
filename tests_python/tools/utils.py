""" Utility functions to check time-dependent assertions in the tests.

Assertions are retried to avoid using arbitrary time constants in test.
"""
from typing import List  # pylint: disable=unused-import
import time
import re
import requests
from client.client import Client
from . import constants


def retry(timeout: float, attempts: float):  # pylint: disable=unused-argument
    """Retries execution of a decorated function until it returns True.

    Args:
        attempts (int): max number of attempts.
        timeout (float): time to wait between attempts.

    Returns:
        True iff an attempt was successful.
    """
    def decorator_retry(func):
        def wrapper(*args, **kwargs):
            nonlocal timeout, attempts
            while not func(*args, **kwargs):
                if attempts == 0:
                    print("*** Failed after too many retries")
                    return False
                print(f'*** Will retry after {timeout} seconds...')
                time.sleep(timeout)
                attempts -= 1
            return True
        return wrapper
    return decorator_retry


@retry(timeout=1., attempts=10)
def check_contains_operations(client: Client,
                              operation_hashes: List[str]) -> bool:
    res = client.rpc('get', '/chains/main/blocks/head/operation_hashes')
    flatten = (res[0] + res[1] + res[2] + res[3] if res is not None and
               len(res) == 4 else [])
    return all(oh in flatten for oh in operation_hashes)


@retry(timeout=1., attempts=20)
def check_protocol(client: Client, proto: str,
                   params: List[str] = None) -> bool:
    res = client.rpc('get', '/chains/main/blocks/head/metadata', params=params)
    return res['next_protocol'] == proto


@retry(timeout=1., attempts=10)
def check_level(client: Client, level) -> bool:
    return client.get_level() == level


@retry(timeout=1., attempts=10)
def check_level_greater_than(client: Client, level) -> bool:
    return client.get_level() >= level


@retry(timeout=2., attempts=20)
def check_operation_in_receipt(client: Client,
                               operation_hash: str,
                               check_previous=None) -> bool:
    extra_param = (['--check-previous', str(check_previous)] if
                   check_previous else [])
    receipt = client.get_receipt(operation_hash, extra_param)
    # TODO deal with case where operation isn't included yet
    return receipt.block_hash is not None


@retry(timeout=5, attempts=20)
def synchronize(clients: List[Client], max_diff: int = 2) -> bool:
    """Return when nodes head levels are within max_diff units"""
    levels = [client.get_level() for client in clients]
    return max(levels) - min(levels) <= max_diff


def get_block_hash(client: Client, level: int) -> str:
    """Return block hash at given level, level must be less or equal
       than current head."""
    cur = 'head'
    while True:
        block = client.rpc('get', f'/chains/main/blocks/{cur}')
        assert level <= block['header']['level']
        if block['header']['level'] == level:
            block_hash = block['hash']
            assert isinstance(block_hash, str)
            return str(block)
        cur = block['header']['predecessor']


def all_blocks(client: Client) -> List[dict]:
    """Return list of all blocks"""
    cur = 'head'
    blocks = []
    while True:
        block = client.rpc('get', f'/chains/main/blocks/{cur}')
        blocks.append(block)
        cur = block['header']['predecessor']
        if block['header']['level'] == 0:
            break
    return list(reversed(blocks))


def operations_hash_from_block(block):
    # TODO type
    _, _, _, operations = block['operations']
    res = []
    for operation in operations:
        res.append(operation['hash'])
    return res


def check_logs(logs: List[str], pattern: str) -> bool:
    for file in logs:
        with open(file, "r") as stream:
            for line in stream:
                if re.search(pattern, line):
                    print('#', stream.name)
                    print(line)
                    return False
    return True


def check_logs_counts(logs: List[str], pattern: str) -> int:
    count = 0
    for file in logs:
        with open(file, "r") as stream:
            for line in stream:
                if re.search(pattern, line):
                    print('#', stream.name)
                    print(line)
                    count += 1
    return count


def activate_alpha(client, parameters=None):
    if parameters is None:
        parameters = constants.PARAMETERS
    proto = constants.ALPHA
    client.activate_protocol_json(proto, parameters)


def rpc(server: str, port: int, verb: str, path: str, data: dict = None,
        headers: dict = None):
    """Calls a REST API

    Simple wrapper over `requests` methods.
    See `https://2.python-requests.org/en/master/`.

    Parameters:
        server (str): server name/IP
        port (int):  server port
        verb (str): 'get', 'post' or 'options'
        path (str): path of the RPC
        data (dict): json data if post method is used
        headers (dicts): optional headers

    Returns:
        A `Response` object."""

    assert verb in {'get', 'post', 'options'}
    full_path = f'http://{server}:{port}/{path}'
    print(f'# calling RPC {verb} {full_path}')
    if verb == 'get':
        res = requests.get(full_path, headers=headers)
    elif verb == 'post':
        res = requests.post(full_path, json=data, headers=headers)
    else:
        res = requests.options(full_path, json=data, headers=headers)
    return res
