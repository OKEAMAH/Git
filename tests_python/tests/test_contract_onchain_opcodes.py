import pytest
from tools import paths
from tools.utils import check_run_failure, assert_storage_contains, bake, \
    init_with_transfer, assert_balance
from tools.constants import IDENTITIES

from client.client import Client

CONTRACT_PATH = f'{paths.TEZOS_HOME}/src/bin_client/test/contracts/opcodes/'
KEY1 = 'foo'
KEY2 = 'bar'


@pytest.mark.incremental
@pytest.mark.slow
@pytest.mark.contract
class TestContractOnchainOpcodes:
    """Tests for individual opcodes that requires origination."""

    def test_gen_keys(self, client):
        """Add keys used by later tests."""
        client.gen_key(KEY1)
        client.gen_key(KEY2)

    def test_store_input(self, client):
        client.transfer(1000, "bootstrap1", KEY1, ['--burn-cap', '0.257'])
        bake(client)

        client.transfer(2000, "bootstrap1", KEY2, ['--burn-cap', '0.257'])
        bake(client)

        assert_balance(client, KEY1, 1000)
        assert_balance(client, KEY2, 2000)

        # Create a contract and transfer 100 ꜩ to it
        init_with_transfer(client, f'{CONTRACT_PATH}/store_input.tz',
                           '""', 100, 'bootstrap1')

        client.transfer(100, "bootstrap1", "store_input",
                        ["-arg", '"abcdefg"', '--burn-cap', '10'])
        bake(client)

        assert_balance(client, "store_input", 200)

        assert_storage_contains(client, "store_input", '"abcdefg"')

        client.transfer(100, "bootstrap1", "store_input",
                        ["-arg", '"xyz"', '--burn-cap', '10'])
        bake(client)

        assert_storage_contains(client, "store_input", '"xyz"')

    def test_transfer_amount(self, client):
        init_with_transfer(client,
                           f'{CONTRACT_PATH}/transfer_amount.tz',
                           '0', 100, 'bootstrap1')

        client.transfer(500, "bootstrap1", 'transfer_amount',
                        ['-arg', 'Unit', '--burn-cap', '10'])
        bake(client)

        assert_storage_contains(client, "transfer_amount", '500000000')

    def test_now(self, client):
        init_with_transfer(client,
                           f'{CONTRACT_PATH}/store_now.tz',
                           '"2017-07-13T09:19:01Z"', 100, 'bootstrap1')

        client.transfer(500, "bootstrap1", 'store_now',
                        ['-arg', 'Unit', '--burn-cap', '10'])
        bake(client)

        assert_storage_contains(client, 'store_now',
                                f'"{client.get_now()}"')

    def test_transfer_tokens(self, client):
        """Tests TRANSFER_TOKENS."""
        client.originate('test_transfer_account1',
                         100,
                         'bootstrap1',
                         f'{CONTRACT_PATH}/noop.tz',
                         ['--burn-cap', '10'])
        bake(client)

        client.originate('test_transfer_account2',
                         20,
                         'bootstrap1',
                         f'{CONTRACT_PATH}/noop.tz',
                         ['--burn-cap', '10'])
        bake(client)

        init_with_transfer(client, f'{CONTRACT_PATH}/transfer_tokens.tz',
                           'Unit', 1000, 'bootstrap1')

        assert_balance(client, 'test_transfer_account1', 100)

        account1_addr = client.get_contract_address('test_transfer_account1')
        client.transfer(100, 'bootstrap1', 'transfer_tokens',
                        ['-arg', f'"{account1_addr}"', '--burn-cap', '10'])
        bake(client)

        # Why isn't this 200 ꜩ? Baking fee?
        assert_balance(client, 'test_transfer_account1', 200)

        account2_addr = client.get_contract_address('test_transfer_account2')
        client.transfer(100, 'bootstrap1', 'transfer_tokens',
                        ['-arg', f'"{account2_addr}"', '--burn-cap', '10'])
        bake(client)

        assert_balance(client, 'test_transfer_account2', 120)

    def test_self(self, client):
        init_with_transfer(client, f'{CONTRACT_PATH}/self.tz',
                           '"tz1KqTpEZ7Yob7QbPE4Hy4Wo8fHG8LhKxZSx"',
                           1000, 'bootstrap1')

        client.transfer(0, 'bootstrap1', 'self', ['--burn-cap', '10'])
        bake(client)

        self_addr = client.get_contract_address('self')
        assert_storage_contains(client, 'self', f'"{self_addr}"')

    def test_contract_fails(self, client):
        init_with_transfer(client, f'{CONTRACT_PATH}/contract.tz',
                           'Unit',
                           1000, 'bootstrap1')

        client.transfer(0, 'bootstrap1', 'self', ['--burn-cap', '10'])
        bake(client)
        addr = client.get_contract_address('contract')

        def cmd():
            client.transfer(
                0, 'bootstrap1', 'contract',
                ['-arg', f'"{addr}"', '--burn-cap', '10'])

        assert check_run_failure(cmd, r'script reached FAILWITH instruction')

    def test_init_proxy(self, client):
        init_with_transfer(client,
                           f'{CONTRACT_PATH}/proxy.tz',
                           'Unit',
                           1000, 'bootstrap1')

    def test_source(self, client):
        init_store = IDENTITIES['bootstrap4']['identity']
        init_with_transfer(client,
                           f'{CONTRACT_PATH}/source.tz',
                           f'"{init_store}"',
                           1000, 'bootstrap1')

        # direct transfer to the contract
        client.transfer(0, 'bootstrap2', 'source', ['--burn-cap', '10'])
        bake(client)

        source_addr = IDENTITIES['bootstrap2']['identity']
        assert_storage_contains(client, 'source', f'"{source_addr}"')

        # indirect transfer to the contract through proxy
        contract_addr = client.get_contract_address('source')
        client.transfer(0, 'bootstrap2', 'proxy',
                        ['--burn-cap', '10', '--arg', f'"{contract_addr}"'])
        bake(client)
        assert_storage_contains(client, 'source', f'"{source_addr}"')

    def test_sender(self, client):
        init_store = IDENTITIES['bootstrap4']['identity']
        init_with_transfer(client,
                           f'{CONTRACT_PATH}/sender.tz',
                           f'"{init_store}"',
                           1000, 'bootstrap1')

        # direct transfer to the contract
        client.transfer(0, 'bootstrap2', 'sender', ['--burn-cap', '10'])
        bake(client)

        sender_addr = IDENTITIES['bootstrap2']['identity']
        assert_storage_contains(client, 'sender', f'"{sender_addr}"')

        # indirect transfer to the contract through proxy
        contract_addr = client.get_contract_address('sender')
        proxy_addr = client.get_contract_address('proxy')
        client.transfer(0, 'bootstrap2', 'proxy',
                        ['--burn-cap', '10', '--arg', f'"{contract_addr}"'])
        bake(client)
        assert_storage_contains(client, 'sender', f'"{proxy_addr}"')

    def test_slice(self, client):
        init_with_transfer(
            client, f'{CONTRACT_PATH}/slices.tz',
            '"sppk7dBPqMPjDjXgKbb5f7V3PuKUrA4Zuwc3c3H7XqQerqPUWbK7Hna"',
            1000, 'bootstrap1')

    @pytest.mark.parametrize('contract_arg',
                             [line.rstrip('\n')
                              for line
                              in open(f'{paths.TEZOS_HOME}/tests_python/tests/'
                                      + 'test_slice_fails_params.txt')])
    def test_slice_fails(self, client, contract_arg):
        def cmd():
            client.transfer(
                0, 'bootstrap1', 'slices',
                ['-arg', contract_arg, '--burn-cap', '10'])

        assert check_run_failure(cmd, r'script reached FAILWITH instruction')
        # bake(client)

    @pytest.mark.parametrize('contract_arg',
                             [line.rstrip('\n')
                              for line
                              in open(f'{paths.TEZOS_HOME}/tests_python/tests/'
                                      + 'test_slice_success_params.txt')])
    def test_slice_success(self, client, contract_arg):
        client.transfer(0, 'bootstrap1', 'slices',
                        ['-arg', contract_arg, '--burn-cap', '10'])
        bake(client)

    def test_split_string(self, client):
        init_with_transfer(client, f'{CONTRACT_PATH}/split_string.tz',
                           '{}',
                           1000, 'bootstrap1')

        client.transfer(0, 'bootstrap1', 'split_string',
                        ['-arg', '"abc"', '--burn-cap', '10'])
        bake(client)
        assert_storage_contains(client, 'split_string',
                                '{ "a" ; "b" ; "c" }')

        client.transfer(0, 'bootstrap1', 'split_string',
                        ['-arg', '"def"', '--burn-cap', '10'])
        bake(client)
        assert_storage_contains(client, 'split_string',
                                '{ "a" ; "b" ; "c" ; "d" ; "e" ; "f" }')

    def test_split_bytes(self, client):
        init_with_transfer(client, f'{CONTRACT_PATH}/split_bytes.tz',
                           '{}',
                           1000, 'bootstrap1')

        client.transfer(0, 'bootstrap1', 'split_bytes',
                        ['-arg', '0xaabbcc', '--burn-cap', '10'])
        bake(client)
        assert_storage_contains(client, 'split_bytes',
                                '{ 0xaa ; 0xbb ; 0xcc }')

        client.transfer(0, 'bootstrap1', 'split_bytes',
                        ['-arg', '0xddeeff', '--burn-cap', '10'])
        bake(client)
        assert_storage_contains(client, 'split_bytes',
                                '{ 0xaa ; 0xbb ; 0xcc ; 0xdd ; 0xee ; 0xff }')

    def test_set_delegate(self, client):
        init_with_transfer(client, f'{CONTRACT_PATH}/set_delegate.tz',
                           'Unit', 1000, 'bootstrap1')
        bake(client)

        assert client.get_delegate('set_delegate').delegate is None

        addr = IDENTITIES['bootstrap5']['identity']
        client.transfer(0, 'bootstrap1', 'set_delegate',
                        ['-arg', f'(Some "{addr}")'])
        bake(client)

        assert client.get_delegate('set_delegate').delegate == addr

        client.transfer(0, 'bootstrap1', 'set_delegate',
                        ['-arg', f'None'])
        bake(client)

        assert client.get_delegate('set_delegate').delegate is None
