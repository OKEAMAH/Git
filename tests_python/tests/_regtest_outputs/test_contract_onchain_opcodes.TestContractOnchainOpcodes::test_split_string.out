tests/test_contract_onchain_opcodes.py::TestContractOnchainOpcodes::test_split_string

Node is bootstrapped.
Estimated gas: 18627 units (will add 100 for safety)
Estimated storage: 574 bytes added (will add 20 for safety)
Operation successfully injected in the node.
Operation hash is '[BLOCK_HASH]'
NOT waiting for the operation to be included.
Use command
  tezos-client wait for [BLOCK_HASH] to be included --confirmations 30 --branch [BLOCK_HASH]
and/or an external block explorer to make sure that it has been included.
This sequence of operations was run:
  Manager signed operations:
    From: [CONTRACT_HASH]
    Fee to the baker: ꜩ0.002422
    Expected counter: [EXPECTED_COUNTER]
    Gas limit: 18727
    Storage limit: 594 bytes
    Balance updates:
      [CONTRACT_HASH] ................ -ꜩ0.002422
      fees(the baker who will include this operation,3) ... +ꜩ0.002422
    Origination:
      From: [CONTRACT_HASH]
      Credit: ꜩ1000
      Script:
        { parameter string ;
          storage (list string) ;
          code { UNPAIR ;
                 DIP { NIL string ; SWAP ; ITER { CONS } } ;
                 DUP ;
                 SIZE ;
                 PUSH nat 0 ;
                 CMPNEQ ;
                 DIP { PUSH @index nat 0 } ;
                 LOOP { PAIR ;
                        DUP ;
                        DIP { UNPAIR ; DIP { PUSH nat 1 } ; SLICE ; ASSERT_SOME ; CONS @storage } ;
                        UNPAIR ;
                        PUSH nat 1 ;
                        ADD @index ;
                        DUP ;
                        DIP { DIP { DUP } ; SWAP ; SIZE ; CMPNEQ } ;
                        SWAP } ;
                 DROP ;
                 DROP ;
                 NIL string ;
                 SWAP ;
                 ITER { CONS } ;
                 NIL operation ;
                 PAIR } }
        Initial storage: {}
        No delegate for this contract
        This origination was successfully applied
        Originated contracts:
          [CONTRACT_HASH]
        Storage size: 317 bytes
        Paid storage size diff: 317 bytes
        Consumed gas: 18627
        Balance updates:
          [CONTRACT_HASH] ... -ꜩ0.317
          [CONTRACT_HASH] ... -ꜩ0.257
          [CONTRACT_HASH] ... -ꜩ1000
          [CONTRACT_HASH] ... +ꜩ1000

New contract [CONTRACT_HASH] originated.
Contract memorized as split_string.
Injected block [BLOCK_HASH]
Node is bootstrapped.
Estimated gas: 17921 units (will add 100 for safety)
Estimated storage: 18 bytes added (will add 20 for safety)
Operation successfully injected in the node.
Operation hash is '[BLOCK_HASH]'
NOT waiting for the operation to be included.
Use command
  tezos-client wait for [BLOCK_HASH] to be included --confirmations 30 --branch [BLOCK_HASH]
and/or an external block explorer to make sure that it has been included.
This sequence of operations was run:
  Manager signed operations:
    From: [CONTRACT_HASH]
    Fee to the baker: ꜩ0.002065
    Expected counter: [EXPECTED_COUNTER]
    Gas limit: 18021
    Storage limit: 38 bytes
    Balance updates:
      [CONTRACT_HASH] ................ -ꜩ0.002065
      fees(the baker who will include this operation,3) ... +ꜩ0.002065
    Transaction:
      Amount: ꜩ0
      From: [CONTRACT_HASH]
      To: [CONTRACT_HASH]
      Parameter: "abc"
      This transaction was successfully applied
      Updated storage: { "a" ; "b" ; "c" }
      Storage size: 335 bytes
      Paid storage size diff: 18 bytes
      Consumed gas: 17921
      Balance updates:
        [CONTRACT_HASH] ... -ꜩ0.018

Injected block [BLOCK_HASH]
{ "a" ; "b" ; "c" }
Node is bootstrapped.
Estimated gas: 18509 units (will add 100 for safety)
Estimated storage: 18 bytes added (will add 20 for safety)
Operation successfully injected in the node.
Operation hash is '[BLOCK_HASH]'
NOT waiting for the operation to be included.
Use command
  tezos-client wait for [BLOCK_HASH] to be included --confirmations 30 --branch [BLOCK_HASH]
and/or an external block explorer to make sure that it has been included.
This sequence of operations was run:
  Manager signed operations:
    From: [CONTRACT_HASH]
    Fee to the baker: ꜩ0.002123
    Expected counter: [EXPECTED_COUNTER]
    Gas limit: 18609
    Storage limit: 38 bytes
    Balance updates:
      [CONTRACT_HASH] ................ -ꜩ0.002123
      fees(the baker who will include this operation,3) ... +ꜩ0.002123
    Transaction:
      Amount: ꜩ0
      From: [CONTRACT_HASH]
      To: [CONTRACT_HASH]
      Parameter: "def"
      This transaction was successfully applied
      Updated storage: { "a" ; "b" ; "c" ; "d" ; "e" ; "f" }
      Storage size: 353 bytes
      Paid storage size diff: 18 bytes
      Consumed gas: 18509
      Balance updates:
        [CONTRACT_HASH] ... -ꜩ0.018

Injected block [BLOCK_HASH]
{ "a" ; "b" ; "c" ; "d" ; "e" ; "f" }
